returns/interfaces/specific/failable.py

```diff
@@ -0,0 +1,165 @@
+        container: 'DiverseFailableN[_FirstType, _SecondType, _ThirdType]',
+        function: Callable[[_FirstType], _NewFirstType],
+    ) -> None:
+        """Ensures that you cannot map a failure."""
+        assert_equal(
+            container.from_failure(raw_value),
+            container.from_failure(raw_value).map(function),
+        )
+
+    @law_definition
+    def bind_short_circuit_law(
+        raw_value: _SecondType,
+        container: 'DiverseFailableN[_FirstType, _SecondType, _ThirdType]',
+        function: Callable[
+            [_FirstType],
+            KindN['DiverseFailableN', _NewFirstType, _SecondType, _ThirdType],
+        ],
+    ) -> None:
+        """
+        Ensures that you cannot bind a failure.
+
+        See: https://wiki.haskell.org/Typeclassopedia#MonadFail
+        """
+        assert_equal(
+            container.from_failure(raw_value),
+            container.from_failure(raw_value).bind(function),
+        )
+
+    @law_definition
+    def alt_short_circuit_law(
+        raw_value: _SecondType,
+        container: 'DiverseFailableN[_FirstType, _SecondType, _ThirdType]',
+        function: Callable[[_SecondType], _NewFirstType],
+    ) -> None:
+        """Ensures that you cannot alt a success."""
+        assert_equal(
+            container.from_value(raw_value),
+            container.from_value(raw_value).alt(function),
+        )
+
+
+class FailableN(
+    ContainerN[_FirstType, _SecondType, _ThirdType],
+    RescuableN[_FirstType, _SecondType, _ThirdType],
+    Lawful['FailableN[_FirstType, _SecondType, _ThirdType]'],
+):
+    """
+    Base type for types that have just only one failed value.
+
+    Like ``Maybe`` types where the only failed value is ``None``.
+    """
+
+    _laws: ClassVar[Sequence[Law]] = (
+        Law3(_FailableLawSpec.rescue_short_circuit_law),
+    )
+
+
+#: Type alias for kinds with two type arguments.
+Failable2 = FailableN[_FirstType, _SecondType, NoReturn]
+
+#: Type alias for kinds with three type arguments.
+Failable3 = FailableN[_FirstType, _SecondType, _ThirdType]
+
+
+class DiverseFailableN(
+    FailableN[_FirstType, _SecondType, _ThirdType],
+    BiMappableN[_FirstType, _SecondType, _ThirdType],
+    Lawful['DiverseFailableN[_FirstType, _SecondType, _ThirdType]'],
+):
+    """
+    Base type for types that have any failed value.
+
+    Like ``Result`` types.
+    """
+
+    _laws: ClassVar[Sequence[Law]] = (
+        Law3(_DiverseFailableLawSpec.map_short_circuit_law),
+        Law3(_DiverseFailableLawSpec.bind_short_circuit_law),
+        Law3(_DiverseFailableLawSpec.alt_short_circuit_law),
+    )
+
+    @abstractmethod
+    def swap(
+        self: _DiverseFailableType,
+    ) -> KindN[_DiverseFailableType, _SecondType, _FirstType, _ThirdType]:
+        """Swaps value and error types in ``DiverseFailableN``."""
+
+    @classmethod
+    @abstractmethod
+    def from_failure(
+        cls: Type[_DiverseFailableType],
+        inner_value: _SecondType,
```

Comment:
This should be `inner_value: _UpdatedType`
Because otherwise we would have type inference bugs.
Because `_SecondType` is a generic-bound type in this context.
And we need a free one.

Reply:
https://github.com/dry-python/returns/blob/master/returns/interfaces/applicative.py#L157

Reply:
Done!